src/Sarif/AnalyzeContextBase.cs

```diff
@@ -16,42 +19,254 @@ public AnalyzeContextBase()
         // candidates to follow this pattern.
-        public IEnumerable<IOption> GetOptions()
+        public virtual IEnumerable<IOption> GetOptions()
         {
-            return new[]
+            return new IOption[]
             {
+                TracesProperty,
+                ThreadsProperty,
+                RecurseProperty,
+                ResultKindsProperty,
+                DataToInsertProperty,
+                FailureLevelsProperty,
+                TimeoutInMillisecondsProperty,
+                TargetFileSpecifiersProperty,
                 MaxFileSizeInKilobytesProperty
             };
         }
 
-        public abstract Uri TargetUri { get; set; }
-        public abstract string MimeType { get; set; }
-        public abstract HashData Hashes { get; set; }
-        public abstract Exception TargetLoadException { get; set; }
-        public abstract bool IsValidAnalysisTarget { get; }
-        public abstract ReportingDescriptor Rule { get; set; }
+        public virtual string PostUri { get; set; }
+        public virtual ISet<string> InvocationPropertiesToLog { get; set; }
+        public virtual ISet<string> PluginFilePaths { get; set; }
+        public virtual ISet<string> PropertiesToLog { get; set; }
+        public virtual ISet<string> InsertProperties { get; set; }
+        public virtual Guid? AutomationGuid { get; set; }
+        public virtual bool Quiet { get; set; }
+        public virtual IFileSystem FileSystem { get; set; }
+        public virtual CancellationToken CancellationToken { get; set; }
+        public virtual IArtifactProvider TargetsProvider { get; set; }
+        public virtual IEnumeratedArtifact CurrentTarget { get; set; }
+        public virtual string MimeType { get; set; }
+        public virtual HashData Hashes { get; set; }
+        public virtual Exception RuntimeException { get; set; }
+        public virtual bool IsValidAnalysisTarget { get; set; }
+        public virtual ReportingDescriptor Rule { get; set; }
         public PropertiesDictionary Policy { get; set; }
-        public abstract IAnalysisLogger Logger { get; set; }
-        public abstract RuntimeConditions RuntimeErrors { get; set; }
-        public abstract bool AnalysisComplete { get; set; }
-        public abstract DefaultTraces Traces { get; set; }
+        public virtual IAnalysisLogger Logger { get; set; }
+        public virtual RuntimeConditions RuntimeErrors { get; set; }
+        public virtual bool AnalysisComplete { get; set; }
+
+        public bool Minify => OutputFileOptions.HasFlag(FilePersistenceOptions.Minify);
+        public bool Optimize => OutputFileOptions.HasFlag(FilePersistenceOptions.Optimize);
+        public bool PrettyPrint => OutputFileOptions.HasFlag(FilePersistenceOptions.PrettyPrint);
+        public bool ForceOverwrite => OutputFileOptions.HasFlag(FilePersistenceOptions.ForceOverwrite);
+
+        public virtual string AutomationId
```

Comment:
>AutomationId [](http://example.com/codeflow?start=30&length=12)

This is the way we create class 'surface area' for properties that are backed by the policy property bag. That dictionary can be kept empty, in which case any retrieval of the static gets the default value from the static descriptor.

This also means that the policy dictionary only contains data values that override defaults, which allows for very compact serialized representation.